course05/homework02/freekpaans/linked_list.h

```diff
@@ -0,0 +1,166 @@
+#include <cstddef>
+#include <memory>
+#include <iterator>
+#include <numeric>
+
+template <typename T>
+class LinkedList {
+public:
+	using size_type = std::size_t;
+	using value_type = T;
+	using reference = T&;
+	using const_reference = const T&;
+
+	struct Link
+	{
+		T mValue;
+		std::unique_ptr<Link> mNext;
+	};
+
+	using LinkPtr = std::unique_ptr<Link>;
+
+	LinkedList() =default;
+	LinkedList(const LinkedList& from);
+
+	LinkedList& operator=(const LinkedList& from);
+
+	bool empty() const
+	{
+		return !mFront;
+	}
+
+	size_type size() const;
+
+	void push_front(const_reference str);
+	void pop_front();
+	void insert_at(size_type index, const_reference str);
+
+	const_reference front() const;
+	const_reference at(size_type index) const;
+
+	struct iterator : public std::iterator<std::input_iterator_tag, T, size_type>
+	{
+		const Link* mLink;
+
+		explicit iterator(const Link* link) : mLink{link}
+		{}
+
+		bool operator==(const iterator& other) const
+		{
+			return mLink == other.mLink;
+		}
+
+		bool operator!=(const iterator& other) const
+		{
+			return !(*this == other);
+		}
+
+		iterator& operator++()
+		{
+			mLink = mLink->mNext.get();
+			return *this;
+		}
+
+		const_reference operator*() const
+		{
+			return mLink->mValue;
+		}
+	};
+
+	iterator begin() const
+	{
+		return iterator{mFront.get()};
+	}
+
+	iterator end() const
+	{
+		return iterator{nullptr};
+	}
+
+private:
+	LinkPtr mFront{};
+};
+
+template <typename T>
+static void push_at(
+	typename LinkedList<T>::LinkPtr& link,
+	typename LinkedList<T>::const_reference item)
+{
+	typename LinkedList<T>::LinkPtr newLink{new typename LinkedList<T>::Link { item, {} }};
+
+	newLink->mNext = std::move(link);
+	link = std::move(newLink);
+}
+
+template <typename T>
+LinkedList<T>::LinkedList(const LinkedList<T>& from)
+{
+	std::accumulate(
+		from.begin(), from.end(),
+		std::ref(mFront),
+		[](LinkPtr& output, const_reference item) -> LinkPtr&
+		{
+			push_at<T>(output, item);
+
+			return output->mNext;
+		});
+}
+
+template <typename T>
+LinkedList<T>& LinkedList<T>::operator=(const LinkedList<T>& from)
+{
+	LinkedList copy{from};
+	mFront = std::move(copy.mFront);
+	return *this;
+}
+
+template <typename T>
+typename LinkedList<T>::size_type LinkedList<T>::size() const
```

Comment:
Here the same problem, but then for the return type of a method. Guess a solution could be defining it in the class definition? I have another one a bit below.